src/rust/engine/src/core.rs

```diff
@@ -238,21 +232,38 @@ impl Key {
 }
 
 ///
-/// A wrapper around a Arc<Handle>
+/// We wrap PyObject (which cannot be cloned without acquiring the GIL) in an Arc in order to avoid
+/// accessing the Gil in many cases.
 ///
-#[derive(Clone, Eq, PartialEq)]
-pub struct Value(Arc<Handle>);
+#[derive(Clone)]
+pub struct Value(Arc<PyObject>);
 
 impl Value {
-  pub fn new(handle: Handle) -> Value {
+  pub fn new(handle: PyObject) -> Value {
     Value(Arc::new(handle))
   }
+
+  // NB: Longer name because overloaded in a few places.
+  pub fn consume_into_py_object(self, py: Python) -> PyObject {
+    match Arc::try_unwrap(self.0) {
+      Ok(handle) => handle,
+      Err(arc_handle) => arc_handle.clone_ref(py),
+    }
+  }
 }
 
+impl PartialEq for Value {
+  fn eq(&self, other: &Value) -> bool {
+    externs::equals(&self.0, &other.0)
+  }
+}
+
+impl Eq for Value {}
```

Comment:
Rust newbie question: I tried googling what this does but I'm not sure what to search in the first place. Could you please point me to docs on what this does or briefly explain?

Reply:
See https://doc.rust-lang.org/std/cmp/trait.Eq.html (in particular, the "if you cannot derive `Eq`" bit.)